APMrover2/mode.h

```diff
@@ -1,8 +1,10 @@
 #pragma once
 
+#include <AP_HAL/AP_HAL.h>
+#include "config.h"
+
 #include <stdint.h>
 #include <GCS_MAVLink/GCS_MAVLink.h>  // for MAV_SEVERITY
-#include "defines.h"
```

Comment:
What's up with the changed includes in this file?

Reply:
Because firstly requested changes broke build so I was forced to fix it somehow.

Reply:
That makes no sense, we don't fix things by randomly changing included headers. What error did you have?

Reply:
this was some days ago, so not clear. Modes in Rover don"t did #include config.h, which cause compilation error "something not a name of class" . I added it, but config.h already includes defines.h so this becomes unneeded. There no need to remove it, it just useless

Reply:
The question is, why do you need this in Rover modes? The new macro is only used in Rover.h (which includes config.h) and in system.cpp (which includes Rover.h).

Reply:
some issues can be fixed easier than to answer "why". Something wrong in order of includes

Reply:
No, no, no. That's not fixing an issue, that's hiding an issue! That's the first step to have bugs.
Please remove this change and get me the exact error, then we can check what is the problem and fix it.

Reply:
in any way, config.h is included from all files both in Plane and Copter, so I don't understand a reason to not do the same on Rover.

Reply:
> in any way, config.h is included from all files both in Plane and Copter, so I don't understand a reason to not do the same on Rover.

We need to reduce dependencies, not increase them.

> Moreover here you can see a wrong place to include defines.h, because such files that not #include anythng should be included first, not after a hard includes. So I think that I just fixed a bug in Rover, not bring a new one

There is no rule saying that about includes, nor is that necessarily a good practice. There is no bug here, this file should include what it needs.

Reply:
The rule should be to avoid unnecessary hardnes - to include all user's setup to all .cpp files. This don't add any dependencies

Reply:
It does add a dependency on the config file! That means that any change in the config file, like the one you are doing, would imply a compile of all cpp files.

Reply:
we definitely have very different coding style... I prefer to include config into all sources to have all settings in single place. Compilation time is not very important thing, especially in case of config change, when I prefer clean rebuild.

But as you wish, reverted.